### ISSUE TITLE
Question about loss calculation

### ISSUE THREAD

**Issue by AIRLegend** [minimaxir/gpt-2-simple#94]
Hi all,

I've been reading the code, trying to understand the implementation, but I have a question about the loss function that possibly is kinda dumb.

Suppose a target text: target="The fox went to the forest fast".
Also suppose an input sample  ```context=[The, fox, went, to, the, forest]```.

The loss is calculated as:

```python
loss = tf.reduce_mean(
        input_tensor=tf.nn.sparse_softmax_cross_entropy_with_logits(
            labels=context[:, 1:], logits=output['logits'][:, :-1]))
```
As far as I know, the goal is to predict the nth token given all the previous ones (i.e. "fast", given the context). However, if we set labels as ```[fox, went, to, the, forest]```,   and the logits as the [0, n-1] tokens, how can the model know the target for the last token (n)?  Aren't we teaching the model to output the padded context and leaving as random the last token?

Sorry for the inconvennience.

**Comment by woctezuma** [comment]
I think the offset is already taken into account, so that:

- `context` is `[the, fox, went, to, the, forest]`
- `output['logits']` correspond to the prediction for `[fox, went, to, the, forest, next_token]`

**Comment by AIRLegend** [comment]
So if input ```context=[the, fox, went, to, the, forest]``` and ```output=[went, to, the, forest, slow]```. I understand that  the loss is the xentropy between ```labels=[went, to the, forest]```and ```outputs[:-1]=[went, to, the, forest]```?. If so, what about the "slow" token? . I can't see how is it implied in the calculation as the last output token it's the real target variable but not used for learning.

Thanks in advance @woctezuma.

**Comment by woctezuma** [comment]
My bad if I was not clear. I wanted to say that:

- `context` is `[the, fox, went, to, the, forest, fast]`,
- `output` corresponds to the prediction for `[fox, went, to, the, forest, fast, ???]`.

So that if one wants to compare the prediction (output) to the ground truth (labels), then one has to remove the last element of output (because there is no ground truth to compare it to), and remove the first token of context (because there is no prediction if there is no "past").

That is my understanding of the offsets in this loss.

**Comment by AIRLegend** [comment]
So the last token in the prediction is not used in the loss function?

**Comment by woctezuma** [comment]
That is my understanding, and the reason would be that there is no ground truth to compare it to.

**Comment by AIRLegend** [comment]
Hmmm I think that I understand it. Thank you for you help @woctezuma.

Closing.
